qiskit/transpiler/passes/basis/basis_translator.py

```diff
@@ -477,53 +477,25 @@ def _basis_search(equiv_lib, source_basis, target_basis):
     if not source_basis:
         return []
 
-    all_gates_in_lib = set()
-
-    graph = rustworkx.PyDiGraph()
-    nodes_to_indices = dict()
-    num_gates_for_rule = dict()
-
-    def lazy_setdefault(key):
-        if key not in nodes_to_indices:
-            nodes_to_indices[key] = graph.add_node(key)
-        return nodes_to_indices[key]
-
-    rcounter = 0  # running sum of the number of equivalence rules in the library.
-    for key in equiv_lib._get_all_keys():
-        target = lazy_setdefault(key)
-        all_gates_in_lib.add(key)
-        for equiv in equiv_lib._get_equivalences(key):
-            sources = {
-                Key(name=instruction.operation.name, num_qubits=len(instruction.qubits))
-                for instruction in equiv.circuit
-            }
-            all_gates_in_lib |= sources
-            edges = [
-                (
-                    lazy_setdefault(source),
-                    target,
-                    {"index": rcounter, "rule": equiv, "source": source},
-                )
-                for source in sources
-            ]
-
-            num_gates_for_rule[rcounter] = len(sources)
-            graph.add_edges_from(edges)
-            rcounter += 1
-
     # This is only neccessary since gates in target basis are currently reported by
     # their names and we need to have in addition the number of qubits they act on.
     target_basis_keys = [
         key
         for gate in target_basis
-        for key in filter(lambda key, name=gate: key.name == name, all_gates_in_lib)
+        for key in filter(
+            lambda key, name=gate: key.name == name, equiv_lib.key_to_node_index.keys()
+        )
     ]
+    graph = equiv_lib.graph
+    vis = BasisSearchVisitor(graph, source_basis, target_basis_keys, equiv_lib.num_gates_for_rule)
 
-    vis = BasisSearchVisitor(graph, source_basis, target_basis_keys, num_gates_for_rule)
     # we add a dummy node and connect it with gates in the target basis.
     # we'll start the search from this dummy node.
-    dummy = graph.add_node("dummy starting node")
-    graph.add_edges_from_no_data([(dummy, nodes_to_indices[key]) for key in target_basis_keys])
+    dummy = graph.add_node({"key": ("dummy starting node", 0)})
```

Comment:
I think that we probably need to remove this node afterwards right? Otherwise it will persist in further use of the equivalence lib. This wasn't an issue before because the graph was constructed on demand.

Reply:
Done in https://github.com/Qiskit/qiskit-terra/pull/8790/commits/837b18558b075631bed822f4c3ba2cee13b98b87.